circuits/update_state_tree.circom

```diff
@@ -1,106 +1,266 @@
+  // Checks to see if we should update the
+  // random leaf (Note: 0 is true, anything non-zero is false)
+  signal valid_signature;
+  valid_signature <-- signature_verifier.valid == 0 ? 0 : 1;
 
-  signature_verifier.from_x <== existing_public_key[0]; // public key x
-  signature_verifier.from_y <== existing_public_key[1]; // public key y
+  signal sufficient_vote_credits;
+  sufficient_vote_credits <-- new_vote_credits > 0 ? 0 : 1;
 
-  signature_verifier.R8x <== decrypted_data.out[message_length]; // sig R8x
-  signature_verifier.R8y <== decrypted_data.out[message_length + 1]; // sig R8x
-  signature_verifier.S <== decrypted_data.out[message_length + 2]; // sig S
+  signal correct_nonce;
+  correct_nonce <-- decrypted_command.out[5] == (state_tree_data[4] + 1) ? 0 : 1;
 
-  for (var i=0; i < message_length; i++) {
-    signature_verifier.preimage[i] <== decrypted_data.out[i];
+  signal valid_state_leaf_index;
+  valid_state_leaf_index <-- decrypted_command.out[0] <= state_tree_max_leaf_index ? 0 : 1;
+
+  signal valid_vote_options_leaf_index;
+  valid_vote_options_leaf_index <-- decrypted_command.out[3] <= vote_options_max_leaf_index ? 0 : 1;
+
+  signal valid_update;
+  valid_update = valid_signature + sufficient_vote_credits + correct_nonce + valid_state_leaf_index + valid_vote_options_leaf_index;
+
+
+  // If any of the above checks return false
+  // We update the random leaf instead
+
+  // Note: valid_update 0 is true, 1 is false
+  signal selected_state_tree_hash;
+  selected_state_tree_hash <-- valid_update == 0 ? new_state_tree_leaf.hash : random_leaf;
+
+  signal selected_state_tree_path_elements[depth];
+  for (var i = 0; i < depth; i++) {
+    selected_state_tree_path_elements[i] <-- valid_update == 0 ? state_tree_path_elements[i] : random_leaf_path_elements[i];
   }
 
-  // **** 4. If signature valid, update leaf **** //
-  component new_state_tree = MerkleTreeUpdate(levels);
-  new_state_tree.leaf <== encrypted_data_hash.hash;
-  for (var i = 0; i < levels; i++) {
-    new_state_tree.path_elements[i] <== state_tree_path_elements[i];
-    new_state_tree.path_index[i] <== state_tree_path_index[i];
+  signal selected_state_tree_path_index[depth];
+  for (var i = 0; i < depth; i++) {
+    selected_state_tree_path_index[i] <-- valid_update == 0 ? state_tree_path_index[i] : random_leaf_path_index[i];
+  }
+
+  // selected_state_tree_hash can only be the random leaf
+  // if the no_op flag is set to true (1)
+  signal no_op_valid_update;
+  no_op_valid_update <-- valid_update == 0 ? new_state_tree_leaf.hash : no_op * random_leaf;
+  
+  selected_state_tree_hash === no_op_valid_update;
+
+  component new_state_tree = MerkleTreeUpdate(depth);
+  new_state_tree.leaf <== selected_state_tree_hash;
+  for (var i = 0; i < depth; i++) {
+    new_state_tree.path_elements[i] <== selected_state_tree_path_elements[i];
+    new_state_tree.path_index[i] <== selected_state_tree_path_index[i];
   }
 
   new_state_tree_root <== new_state_tree.root;
```

Comment:
I'd suggest moving this line to the end of the circuit, since it's an output and it feels more intuitive that it would be located there there, like a return statement in a function would be.